haystack/nodes/retriever/link_content.py

```diff
@@ -0,0 +1,197 @@
+import logging
+from datetime import datetime
+from http import HTTPStatus
+from typing import Optional, Dict, List, Union, Callable
+from urllib.parse import urlparse
+
+import requests
+from boilerpy3 import extractors
+from requests import Response
+from requests.exceptions import InvalidURL
+
+from haystack import Document, __version__
+from haystack.document_stores import BaseDocumentStore
+from haystack.nodes import PreProcessor
+from haystack.nodes.retriever.base import BaseRetriever
+from haystack.schema import FilterType
+
+logger = logging.getLogger(__name__)
+
+
+def html_content_handler(response: Response) -> Optional[str]:
+    """
+    Extracts content from the response text using the boilerpy3 extractor.
+    """
+    extractor = extractors.ArticleExtractor(raise_on_failure=False)
+    try:
+        extracted_content = extractor.get_content(response.text)
+        return extracted_content
+    except Exception:
+        return None
+
+
+def pdf_content_handler(response: Response) -> Optional[str]:
+    # TODO: implement this
+    return None
+
+
+class LinkContentRetriever(BaseRetriever):
+    """
+    LinkContentRetriever fetches content from a URL and converts it into a list of Document objects.
+
+    LinkContentRetriever supports the following content types:
+    - HTML
+
+    """
+
+    def __init__(self, pre_processor: Optional[PreProcessor] = None):
+        """
+        Creates a LinkContentRetriever instance.
+        :param pre_processor: PreProcessor to apply to the extracted text
+        """
+        super().__init__()
+        self.pre_processor = pre_processor
+        self.content_handlers: Dict[str, Callable] = {"html": html_content_handler, "pdf": pdf_content_handler}
+
+    def __call__(self, url: str, timeout: int = 3, doc_kwargs: Optional[dict] = None) -> List[Document]:
+        """
```

Comment:
Using the `__call__` method in a Node does not particularly appeal to me. I would prefer a more explicit method name.
(Looking at the codebase, the `PromptNode` seems one of the rare cases where we use this method).

Reply:
Deal, fetch? I'll first read https://benhoyt.com/writings/short-names/ :-)